### PR TITLE
Fix #1453 pointer lost in Swig API

### LINKED ISSUE #1453: [SWIG] Argument pointer destroyed too early in xccdf_session_set_rule
#### Description of Problem:
Hi, I've encountered recently a serious bug in python bindings.
Example:

```
import openscap_api as oscap
sess = oscap.xccdf.session_new("./samples/xccdf_sample.xml")
rule="R-SHOULD_PASS"
sess.set_rule(rule)
sess.load() # should return 0 but return 1
sess.evaluate()
```
After an horrible debugging i've found that when xccdf_session_set_rule is called, the rule is well stored, but when xccdf_session_load is called, the pointer is lost (random/inconsistant value). I've modify xccdf_session_set_rule by adding a malloc/strcpy before store the value in session->rule, and the problem doesn't exists anymore.

Of course it's not a good solution. The best compromise that i've found is to add a wrap function in file openscap.i like

```
void xccdf_session_set_rule_py(struct xccdf_session  *sess, const char *rule) {
    char *n_rule=(char*)malloc((strlen(rule)+1)*sizeof(char));
    xccdf_session_set_rule(sess, rule);
}
```
I've tested that an it works (to avoid memory leak we should also wrapp the xccdf_session_destroy function in orde to release the pointer). I write this as an issue because it's not a trivial problem, and it can probably happens with other functions (i guess that the char* agrument is released when the call to set_rule is terminated ... ).

So, what o you suggest in order to mitigate these problems ? I will propose some wrapping functions in openscap.i in order to have a working API but it's not the better mitigation ever.

What do you think about these kinds of problems concerning SWIG bindings ?

#### OpenSCAP Version:
maint-1.3

#### Operating System & Version:
Ubuntu 18.04

### PR DESCRIPTION
**Pull request by DominiqueDevinci**
Hello,
Some time ago i've signalled a serious issue #1453  in the python binding, the xccdf_session_set_rule was just not working.

Here is a proposal of correction, including two more unit tests which ensure it's working now.
I can understand some technical choices are debatable in this solution but i've to be as clear as possible. If you have a better proposal i'm all ears !

Regards,
Dominique.

### CODE REVIEW

File: swig/openscap.i [hunk 1]
```diff
@@ -549,5 +549,25 @@ int oval_agent_eval_system_py(oval_agent_session_t * asess, PyObject * func, PyO
...
+* and so it's not possible to access to the rule property without defining it
+*/
+struct xccdf_session {
+        char *filename;
+        char *rule;
+};
+
+void xccdf_session_set_rule_py(struct xccdf_session  *sess, char *rule) {
+    char *n_rule=(char*)malloc((strlen(rule)+1)*sizeof(char));
+    strcpy(n_rule, rule);
```
Comment (jan-cerny) [inline swig/openscap.i]: Use `strdup` instead.

File: swig/openscap_api.py [hunk 2]
```diff
@@ -273,13 +273,24 @@ def __getattr__(self, name):
         if name in self.__dict__:
             return self.__dict__[name]
 
+        # cache self.object + "_" + name for perf / clearness purpose
+        obj_name = self.object + "_" + name
+
+        """ Catch  potential C function overriden in OSCAP_Object
+        like xccdf_session_set_rule ir xccdf_session_free for instance. """
```
Comment (jan-cerny) [inline swig/openscap_api.py]: ir should be or?

**Comment by DominiqueDevinci** [comment]
Yes you are right, fixed :)

**Comment by jan-cerny** [comment]
Thanks for your contribution.

**Pull request merged**